### ISSUE TITLE
Why not just bundle react with everything else?

### ISSUE THREAD

**Issue by k88hudson** [MozillaFoundation/mofo-example-app#17]
I have to say I don't really agree with the reasons given here for not bundling React with the rest of the JS code: https://github.com/MozillaFoundation/mofo-example-app/blob/master/Gulpfile.js#L12

There are a lot written about using a CDN v.s. bundling/minifying. The general conclusion seems (to me) to be there is not much difference, however, I would like to argue against using the CDN version since:

* it depends on a working network connection
* it doesn't allow the app to "packaged" for cordova or offline use (without the use of an appcache manifest, see next point)
* it complicates CSP, appcache manifest, and mobile manifest configuration
* it complicates browserify/build tool configuration
* it complicates dependency version updating (i.e. we will have to watch both package.json AND script links in an html file)

Note: @Pomax you mentioned that React browserified is considerably larger when browserified; however, after running some local tests I was unable to reproduce this (the difference I saw was about ~7b unminified). If you are still seeing that issue, it is certainly something we'd need to resolve.

**Comment by davidascher** [comment]
Good points

**Comment by Pomax** [comment]
I literally don't care one way or another. If you think bundling a 130kb minifeid library into every app, fine. If we want to save that by having the browser do cache resolution for the same URL that all our apps will use, also fine. Call it and we'll run with it.

**Comment by davidascher** [comment]
P1 is we need the app to work post-install without network access on first
start.  Sounds like bundling makes that easy, and we'd need to do something
else if we weren't bundling.

**Comment by Pomax** [comment]
if we own the CDN, we can simply add our own link to the appCache manifest and it'll be cached, as far as I know. I'm not saying we shouldn't build full bundles, but I am definitely saying that offline caching isn't an argument for or against bundling. We can do offline caching with either approach =)

(http://www.html5rocks.com/en/tutorials/appcache/beginner/ gives an example with both local and cdn files as one of the very first examples)

**Comment by davidascher** [comment]
Cool, then I don't care about which approach is used.

**Comment by k88hudson** [comment]
That would still require network access for when the app is first loaded (appcache only caches *after* the first time users load the page, not after a download/install)

Also, on a related note, IMO we should be *very* careful when deciding to use appcache for anything, see http://alistapart.com/article/application-cache-is-a-douchebag, esp. when there are other options available. There are *many* strange behaviours that have to be worked around.

**Comment by Pomax** [comment]
Certainly true.

Is there a reason we can't do both? We have control of the build process, so I'm worried about bike shedding: is there a reason we can't generate both a full offline bundle with "everything in it" for platforms where bundles live in isolation of each other, potentially without network access forever, as well as a web bundle that takes advantage of browsers being able to use cache as dedupe mechanism?

Most of the build process will be identical for these two targets anyway, we'd simply be clearer in which parts of the build can be optimized by splitting it off for one of the two deployment targets, using the exact same source files, letting us efficiently target both "the web" and "offline phones" for environments where internet access is a commodity

**Comment by davidascher** [comment]
We should probably decide what platforms this sample app repo is targeting.
E.g. whether taking the build stage all the way to apk generation is in
scope for this repo or another.

**Comment by Pomax** [comment]
that's an excellent point. At this point it almost feels like it's worth setting up a MozillaFoundationAppamples account with several repositories for the various apps (APIs vs. universal libraries vs. client-side web apps vs. universal apps with ios/android/ffos builds). @simonwex any opinions on whether that's worth doing or whether that's getting a little too crazy?

(we're going to have to split this out into several repos already anyway to unmuddle the API vs. client-sided-ness)

**Comment by davidascher** [comment]
Two thoughts about the last q:

a) Want to make sure everyone's on board with the New Way before we start
having copies of it to maintain. Let's get everyone to make a sample app
once some of the current bugs are resolved.

b) In general, having multiple apps that demonstrate The Way makes it
harder to evolve The Way

My preference today would be:
 - an API-only repo
 - an app repo, with optional apk-generation step.  That's important
because it's obviously not ok for the webapp-way to not be adaptable to the
apk-generating scenario for one of our major deliverables this year.
 - an obvious and documented way to do an app that does both, without its
own copy of either code.

Given that, get at least most devs to spend a couple of hours playing w/
that and react to it. (pun intended).

Then iterate as needed.

**Comment by Pomax** [comment]
Fully agreed on fragmentation slowing down progress, I'm also wondering if there's a way we can maybe do this without splitting up the repo at all, by having an `examples` dir with an `api-server` and two client side `gallery` + `editor` dirs, but keep the top level node modules, package.json, gulpfile etc. so we don't end up having to split those up.

I like the idea of fragmentation because it's clean separation of responsibility, but I'd also like to find a way to make sure we keep our build chain etc. limited to a single set of master files, and for that even splitting up into two repos is going to pose problems.
